File: python/euler/bin/p0085.py
# project euler: problem 85

#   nCr = n! / ((n-r)! * r!)
#
#       1  2       n-1  n
#     +--+--+-- ... --+--+
#    1|  |  |   ...   |  |
#     +--+--+-- ... --+--+
#    2|  |  |   ...   |  |
#     +--+--+-- ... --+--+ num of horizontal lines = m + 1
#    3|  |  |   ...   |  |
#     +--+--+-- ... --+--+
#     ....................
#     +--+--+-- ... --+--+
#    m|  |  |   ...   |  |
#     +--+--+-- ... --+--+
#       num of vertical lines = n + 1
#
#   (m+1)C2 * (n+1)C2 = m(m+1)/2 * n(n+1)/2 (\approx) 2_000_000
#   --> m(m+1)*n(n+1) (\approx) 8_000_000

from itertools import count
from math import isqrt

from euler.lib.util import HeapQueue


def get_diff(m: int, target: int) -> tuple[int, int] | None:
    def lhs(m: int, n: int) -> int:
        return m * (m + 1) * n * (n + 1)

    n = isqrt(target // (m * (m + 1))) - 1
    while lhs(m, n) < target:
        n += 1

    if m >= n:
        return None

    if abs(target - lhs(m, n - 1)) < abs(target - lhs(m, n)):
        return abs(target - lhs(m, n - 1)), n - 1
    else:
        return abs(target - lhs(m, n)), n


def compute(target: int) -> str:
    pq = HeapQueue()
    new_target = target * 4
    for m in count(1):
        if (tpl := get_diff(m, new_target)) is None:
            break
        pq.insert((tpl[0], m * tpl[1]))

    return str(pq.peek()[1])


def solve() -> str:
    return compute(2_000_000)

File: typescript/euler/bin/p0085.ts
// project euler: problem 85

/*
  nCr = n! / ((n-r)! * r!)

      1  2       n-1  n
    +--+--+-- ... --+--+
   1|  |  |   ...   |  |
    +--+--+-- ... --+--+
   2|  |  |   ...   |  |
    +--+--+-- ... --+--+ num of horizontal lines = m + 1
   3|  |  |   ...   |  |
    +--+--+-- ... --+--+
    ....................
    +--+--+-- ... --+--+
   m|  |  |   ...   |  |
    +--+--+-- ... --+--+
      num of vertical lines = n + 1

  (m+1)C2 * (n+1)C2 = m(m+1)/2 * n(n+1)/2 (\approx) 2_000_000
  --> m(m+1)*n(n+1) (\approx) 8_000_000
*/

import { ascend, BinaryHeap } from "@std/data-structures";
import { isqrt } from "../lib/math.ts";

const getDiff = (m: number, target: number): [number, number] | undefined => {
  const abs = Math.abs;

  const lhs = (m: number, n: number): number => m * (m + 1) * n * (n + 1);

  let n = isqrt(Math.trunc(target / (m * (m + 1)))) - 1;
  while (lhs(m, n) < target) {
    n += 1;
  }

  if (m >= n) {
    return undefined;
  }
  const r1 = abs(target - lhs(m, n - 1));
  const r2 = abs(target - lhs(m, n));
  if (r1 < r2) {
    return [r1, n - 1];
  } else {
    return [r2, n];
  }
};

export const compute = (target: number): string => {
  const pq = new BinaryHeap<[number, number]>((a, b) => ascend(a[0], b[0]));
  const new_target = target * 4;
  let m = 0;
  while (true) {
    m += 1;
    const result = getDiff(m, new_target);
    if (result === undefined) {
      break;
    }
    pq.push([result[0], m * result[1]]);
  }

  return String(pq.peek()![1]);
};

export const solve = (): string => compute(2_000_000);
